### ISSUE TITLE
maxDuration not being respected

### ISSUE THREAD

**Issue by jimthedev** [CharlesStover/fetch-suspense#1]
Hi @CharlesStover! First off thanks for writing this package.

The only thing that I noticed was that maxDuration seems to not be respected and the loading spinner seems to show right away. Have you experienced this at all?

**Comment by quisido** [comment]
I have not tested it with maxDuration, but that sounds like it would be an issue with the Suspense component. The Suspense component does not tell the children components how long they need to wait before throwing a Promise. Suspense should be what does the waiting after catching a Promise before re-rendering to a fallback.

Is it possible that your version of React is pre-`maxDuration`, when Suspense called that prop `delayMs` instead?

**Comment by jimthedev** [comment]
Good question. I created a codesandbox to show the issue:

https://codesandbox.io/s/38692wo8x5

To reproduce make sure you:

1. open the codesandbox
1. open your dev tools and change over to slow 3g
1. click the load it button.

I would expect that the loading spinner no show up for at least two seconds since my suspense element has a 2000 ms maxDuration but instead it shows up immediately.

I am happy to try other things but hopefully this gives you a starting point.

**Comment by jimthedev** [comment]
@CharlesStover The more I look at it I don't think fetch-suspense is actually being caught by suspense at all so this might not have anything to do with maxDuration but with the core functionality of the library. Do you have an example of this library actually working with suspense?

**Comment by quisido** [comment]
The code in the repo definitely works with Suspense, as it was tested before being pushed to production. Dan Abramov himself actually checked out this repo soon after release. He commented on how the promise memoization would cause a memory leak without cache invalidation (which is why there is a third argument for the hook), but didn't have any problems with the conceptual interaction between Suspense and the hook. If for some reason the package doesn't work, there may have been an error when building the production package.

Unfortunately, CodeSandbox is blocked by my workplace firewall, so I am unable to see your code at this moment.

I will try building a quick demo using the live NPM package instead of my local production build to see if it behaves differently and if I can correctly emulate Suspense's behavior.

**Comment by quisido** [comment]
I’ve updated my [go-to demo project](https://mtgenius.github.io/proxy/) to use Suspense + fetch-suspense for its API calls. You can see the use of useFetch and React Suspense components in the following files:

[search.js wraps the Results component with Suspense.](https://github.com/mtgenius/proxy/blob/master/src/components/search/search.js#L24)
[The Results component calls useFetch when its value prop does not exist in the global state.](https://github.com/mtgenius/proxy/blob/master/src/components/search/results/search-results.js#L16)

The live demo may not be intuitive to use, because it was designed for private use and does not implement optimal UI/UX, but here is a visual walkthrough above the aforementioned components and behavior:

![image](https://user-images.githubusercontent.com/343837/48570057-fcf7c680-e8c8-11e8-97d6-ea2be69982a9.png)

Hope this helps!

**Comment by jimthedev** [comment]
Ok gotcha so based on that the Suspense feature of grouping siblings and ancestor fetches works. It truly is just the maxDuration that currently isn't being obeyed.

**Comment by quisido** [comment]
I tried `maxDuration` in the above project, and I agree it looks like it is not being respected. I think this would be a problem with Suspense, because -- unless Suspense has a context -- I don't believe is possible for the hook to know what the `maxDuration` is. That is on Suspense to handle.

That said, if they were to release a patch for this, it would likely come out in the same version that they integrate the Suspense hook, making it a `wontfix` as they would encourage use of the Suspense hook instead. This package is meant to bridge the gap between 16.6 and the Suspense hook, and it looks like it will simply be a limitation that `maxDuration` cannot be supported for that version gap.

If I am wrong and Suspense is using a context for `maxDuration`, I'll be more than happy to update the package. I'm fairly positive that it is not.

**Comment by RobinMalfait** [comment]
Max duration is currently not implemented in the current suspense component. Currently this is expected behaviour
